lambda/translcore.ml

```diff
@@ -322,36 +329,155 @@ let can_apply_primitive p pmode pos args =
+                kind)
+
+(* Translates lists of expressions, which must have at least one non-void
+   element.  The resulting list has an element for each non-void element of the
+   input.  Computations corresponding to the void elements are attached to the
+   non-void elements in a way that preserves evaluation order.
+
+   This is generalized to support using it on both lists of expressions and
+   lists of record fields.  So it takes an ['a list] and some functions to
+   do the various bits:
+
+   is_void : 'a -> bool
+   value_kind : 'a -> value_kind    (will only be called if not is_void)
+   transl : 'a -> int option -> Lambda.t
+
+*)
+let transl_list_with_voids ~is_void ~value_kind ~transl expr_list =
+  (* before and after are lists of void computations in reverse eval order.
+     If before = after = [], this is the same as just translating e *)
+  let transl_with_voids_before_and_after (before,e,after) =
+    let transl_with_voids_after e after =
+      let kind = value_kind e in
+      match after with
+      | [] -> (kind, transl None e)
+      | _ ->
+        let result_var = Ident.create_local "after_voids" in
+        (kind,
+         Llet (Strict, kind, result_var,
+               transl None e,
+               List.fold_left (fun lam e ->
+                 catch_void (fun void_k -> transl void_k e) lam kind)
+                 (Lvar result_var) after))
+    in
+    let transl_with_voids_before kind lam before =
+      List.fold_left (fun lam e -> catch_void (fun void_k -> transl void_k e)
+                                     lam kind)
+        lam before
+    in
+    let (kind, lam) = transl_with_voids_after e after in
+    (transl_with_voids_before kind lam before, kind)
+  in
+  let group_voids exprs =
+    (* We need to group the voids with an element before them or the element
+       after them.  We prefer to do them as part of an element after them in
+       eval order, because doing them "before" an element is free while the
+       alternative adds a let binding. *)
+    let rec get_afters idx afters = function
+      | e :: exprs when is_void e ->
+        get_afters (idx + 1) (e :: afters) exprs
+      | exprs -> (idx,exprs,afters)
+    in
+    let rec group idx befores nonvoid afters acc exprs =
+      match exprs with
+      | [] -> (List.rev befores,nonvoid, List.rev afters) :: acc
+      | e :: exprs when is_void e ->
+        group (idx+1) (e :: befores) nonvoid afters acc exprs
+      | e :: exprs ->
+        group (idx+1) [] e []
+          ((List.rev befores,nonvoid,List.rev afters) :: acc) exprs
+    in
+    let (idx,exprs,afters) = get_afters 0 [] exprs in
+    let (first_non_void,exprs) = match exprs with
+      | [] -> assert false
+      | (e :: exprs) -> e,exprs
+    in
+    List.rev (group (idx+1) [] first_non_void afters [] exprs)
+  in
+  let grouped = group_voids expr_list in
+  List.split (List.map transl_with_voids_before_and_after grouped)
+
+let rec transl_exp ~scopes void_k e =
```

Comment:
We want to avoid calling type-checking functions from transl, and have set things up so it's otherwise not necessary (it's late to be running them, requiring the use of the potentially expensive `Ctype.correct_levels`).  Also my impression is we prefer to not to add this kind of sanity checking assert in general.

That said, it's still on the TODO list to add more sanity checking asserts before we ship v0, so I'll leave a CR here.  My guess is that recomputing the layout of the type of every term we call `transl_exp` on is prohibitively expensive, and we'll want to do it somewhere else, though.